File: apps/web-svelte/src/lib/api/nav-layout.ts
import { fetchApi } from "$lib/api/orval-fetch";
import { normalizeNavPrefs, type NavPrefs } from "$lib/nav/nav-catalog";

/**
 * Server-persisted navigation layout. Mirrors the `.NET` `NavLayoutDocument` contract;
 * `version` corresponds to {@link NavPrefs.v}. The layout is shared across a user's
 * devices — section grouping, order, hidden items, and the mobile dock all live here.
 */
interface NavLayoutDocument {
  version: number;
  sections: NavPrefs["sections"];
  hidden: string[];
  mobileFavorites: string[];
}

interface NavLayoutResponse {
  /** The stored layout, or null when none has been saved (client uses seeded defaults). */
  layout: NavLayoutDocument | null;
}

function toDocument(prefs: NavPrefs): NavLayoutDocument {
  return {
    version: prefs.v,
    sections: prefs.sections,
    hidden: prefs.hidden,
    mobileFavorites: prefs.mobileFavorites,
  };
}

/**
 * Fetch the server-persisted navigation layout. Returns `null` when nothing is stored
 * or the stored document is malformed, so the caller keeps its seeded defaults.
 */
export async function fetchNavLayout(signal?: AbortSignal): Promise<NavPrefs | null> {
  const response = await fetchApi<NavLayoutResponse>("/nav-layout", { signal });
  if (!response?.layout) return null;
  return normalizeNavPrefs(response.layout);
}

/** Persist the navigation layout, replacing any previously stored layout. */
export async function saveNavLayout(prefs: NavPrefs): Promise<void> {
  await fetchApi<NavLayoutResponse>("/nav-layout", {
    method: "PUT",
    body: JSON.stringify(toDocument(prefs)),
  });
}

File: apps/backend/src/Prismedia.Contracts/Navigation/NavLayout.cs
namespace Prismedia.Contracts.Navigation;

/// <summary>
/// One user-defined navigation section in display order.
/// </summary>
/// <param name="Id">Stable section identity assigned by the client.</param>
/// <param name="Label">User-facing section label.</param>
/// <param name="Items">Ordered route hrefs that belong to this section.</param>
/// <param name="Collapsed">Whether the section renders collapsed in the expanded sidebar.</param>
public sealed record NavLayoutSection(
    string Id,
    string Label,
    IReadOnlyList<string> Items,
    bool Collapsed);

/// <summary>
/// The complete, server-persisted navigation layout shared across a user's devices.
/// Sections, order, and hidden items are common to mobile and desktop; the mobile
/// dock (<see cref="MobileFavorites"/>) only renders on mobile but is stored here too.
/// The backend persists this document opaquely — route hrefs are reconciled against
/// the live client-side catalog at render time, so no server-side href validation is done.
/// </summary>
/// <param name="Version">Schema version of the layout document.</param>
/// <param name="Sections">Sections in display order.</param>
/// <param name="Hidden">Route hrefs hidden from normal (non-edit) rendering.</param>
/// <param name="MobileFavorites">Ordered hrefs shown in the mobile bottom dock.</param>
public sealed record NavLayoutDocument(
    int Version,
    IReadOnlyList<NavLayoutSection> Sections,
    IReadOnlyList<string> Hidden,
    IReadOnlyList<string> MobileFavorites);

/// <summary>
/// Response for the navigation layout endpoints.
/// </summary>
/// <param name="Layout">
/// The stored layout, or <c>null</c> when none has been saved yet — in which case the
/// client falls back to its seeded default layout.
/// </param>
public sealed record NavLayoutResponse(NavLayoutDocument? Layout);
